### PR TITLE
Fix Possible Race Condition in TSDB

### PR DESCRIPTION
**Pull request by mneverov**
Fixes: #7571

Signed-off-by: Max Neverov <neverov.max@gmail.com>

### CODE REVIEW

**Comment by mneverov** [comment]
@codesome could you please take a look?

**Review by codesome — COMMENTED** [review]
We missed 1 thing, we need to move https://github.com/prometheus/prometheus/blob/ca6f2bde9489d274719fcc17069f6a7aabb833b0/tsdb/chunks/head_chunks.go#L357-L363 inside the `cdm.readPathMtx.Lock()` right below. ~LGTM on the remaining part.~

**Comment by codesome** [comment]
We have 2.21 rc going out today, lets try to get this in before that :)

cc @roidelapluie

File: tsdb/chunks/head_chunks.go [hunk 1]
```diff
@@ -720,17 +720,20 @@ func (cdm *ChunkDiskMapper) DeleteCorrupted(originalErr error) error {
 
 	// Delete all the head chunk files following the corrupt head chunk file.
 	segs := []int{}
+	cdm.readPathMtx.Lock()
```
Comment (codesome) [inline tsdb/chunks/head_chunks.go:723]: We should be holding the `RLock()` and `RUnlock()` here.

Comment (mneverov) [reply tsdb/chunks/head_chunks.go]: Fixed

**Review by codesome — APPROVED** [review]
LGTM, thanks! Will merge after test is green.

**Pull request merged**